README.md
# **python_projects**
Last update: 2022-10-15 01:37

<br>
Make sure to update the Projects table: [All Projects Table](https://github.com/users/henryj320/projects/5)
<br><br>

## Future Project Ideas

---
<br>


## **watt_checker**<br>
**Language**: Python / Powershell<br>
**Overview**: Records the Watts-used by the laptop every 30 minutes. Record the value into an Excel spreadsheet so that a chart can be automatically generated or updated every day. Alternatively, use Python so that matplotlib and seaborn can be used. On brief research, it is not easy to grab the current power draw of a computer. Intel has some built-in, but may be complicated.<br><br>


## **weekly_goals_GUI**<br>
**Language**: Python<br>
**Overview**: Shows a progress bar for your weekly goals. For example, a progress bar on how close to your reading goal you are. You could also autorun this on startup and set your background to the progress bars by screenshotting and setting as the wallpaper. It could also show how much of your 7.5 * 5 hrs of work you've done so far. You could run this via Flask. This would make it so that you can use Python (storing values in .txt files) and a clean webpage with Bootstrap. Alternatively, you could use React for clean elements and running.<br><br>


## **wifi_users_tracker**<br>
**Language**: Linux Script<br>
**Overview**: A software which states when a new device joins your WiFi network via a notification. It could be extended to show when a device disconnects. Previous implementation of this used a for loop to read the nmap's "x devices found" string.<br><br>


## **transcripter**<br>
**Language**: Python<br>
**Overview**: Learn how to make a transcripter that converts speech to text. Look into any existing libraries that allow for this.<br><br>


## **internet_stability_recorder**<br>
**Language**: Python<br>
**Overview**: Sends a ping every 5 minutes and records the result to Excel. Use the results to automatically plot into a chart. This would show how stable the internet is throughout the day.<br><br>


## **centre_country_shower**<br>
**Language**: Python<br>
**Overview**: Shows the centre of a given country based on different factors: geological, average age, population, etc. This would show age densities, etc.<br><br>


## **running_route_generator**<br>
**Language**: Python<br>
**Overview**: The user provides a location. The script generates a path to coordinates x metres away and adds this as a route on Google Maps and outputs this to the user. This allows the user to run to a random place every time the exact distance away that they want.<br><br>


## **tcp_message_sender_and_receiver**<br>
**Language**: Python<br>
**Overview**: Write a program that sends and receives message from the basics. Generate a TCP segment with all of the fields in binary. Set up a port to send on and a port to receive on. Use your Java code as a reference. Importantly, the receiver needs to be listening on the port. Use datasubs to find how to run an app on a certain portand listen to requests. Set out the plan logically.<br><br>


## **height_determination_software**<br>
**Language**: Not a clue<br>
**Overview**: Use vanishing points to determine the height of individuals or objects inside photos. (Fspy)<br><br>


## **blender_addon_forensic_tool**<br>
**Language**: Python <br>
**Overview**: Creating simulations, blah, blah, blah.<br><br>


## **computer_writing_tool**<br>
**Language**: Not a clue<br>
**Overview**: Piece of software that analyses text, categorises words into groups (verbs, nouns, prepositions, etc.) and uses preset knowledge on sentence structures to write sentences/more.<br><br>


## **dnd_map_precedural_gen_tool**<br>
**Language**: Not a clue <br>
**Overview**: Precedurally generates tools in different settings for use by DMs and GMs in TTRPG campaigns. Different settings can be changed to tweak results. <br><br>


## **text_to_music**<br>
**Language**: not a clue <br>
**Overview**: Converts text to music based on a formula. Including tone reading through negativity/positivity of words in sentence, (minor/major). https://www.youtube.com/watch?v=PVT6WpOyf0s&feature=youtu.be<br><br>


## **precedurally_generated_evolution_of_civilisation**<br>
**Language**: Not a clue<br>
**Overview**: Tweak parameters as the civilisation and planet evolves to create stats and planet image.<br><br>


## **dnd_fun_tool**<br>
**Language**: Not a clue <br>
**Overview**: Keep track of characters, better combat system, better image display system, better tools for DM.<br><br>


## **Free_Worlds_website**<br>
**Language**: HTML/Javascript<br>
**Overview**: website go brr.<br><br>


## **full_system_search**<br>
**Language**: Python<br>
**Overview**: Similar to DocumentSystem to Git. Make a program that takes a searchterm and searches the entire computer for any documents named or containing the term. Make it show sections at a time with a progress bar to learn Python progress bars in the terminal. Add a parameter for OS so that OS files can be ignored. Output should be something like "Found: X occurrences of "example" in Documents, Y occurences of "example" in Pictures.<br><br>


## **react_dashboard**<br>
**Language**: React<br>
**Overview**: Develop a React frontend using a [template](https://www.creative-tim.com/templates/react-free). Develop it from scratch making use of a template's components when required. Have separate pages for each Python project. Using axios like you did in the devops sprint to connect to a running API and output the results.

<br><br>


<br>

## Completed Project Ideas

---
<br>


## **python_template**<br>
**Date Started**: 2022-09-15<br>
**Date Completed**: 2022-09-15 (1 day)<br>
**Language**: Python<br>
**Overview**: A template for later Python projects. Containing:
- cookiecutter - A template for the whole project.
- tox - So that we can use pydocstyle and pycodestyle.
- .env - For environment variables.
- test folder - To store tests.
- README
- Example .py module  - Make it reference an environment variable and have a example method that can be run (and how to run it).
- .gitignore - Git has default ones.<br>

**Result**: I now have a cookiecutter template up and running for future Python projects.<br>
**Running the Project**: You can call the cookiecutter template by downloading it to your machine and running:
- "cd...".
- "cookiecutter ~/Documents/python_projects/python_template". <br><br>


## **basic_messaging**<br>
**Date Started**: 2022-09-17<br>
**Date Completed**: 2022-09-17 (1 day)<br>
**Language**: Python<br>
**Overview**: A basic messaging app. Effectively a prerequisite for tcp_message_sender_and_receiver. Use Flask to send a string from one endpoint (host:port pair) to another Python script.<br>
**Result**: You can now send and receive a TCP or UDP message between a server (receiver) and client (sender).
**Running the Project**: You can run the code in the terminal by running:
- Terminal 1: "python receiver.py udp".
- Terminal 2: "python sender.py udp".

You may need to update the host and port.<br><br>


## **muscle_checker**<br>
**Date Started**: 2022-09-29<br>
**Date Completed**: 2022-10-09 (10 days)<br>
**Language**: Python<br>
**Overview**: Input which exercises you completed this week. The system outputs which of the main muscles you missed this week and what exercises you could do to train them.<br>
**Result**: I now have a muscle_checker Flask website which can be run anytime and outputs the muscles hit, muscles missed and what exercises to do to hit them.<br>
**Running the Project**: To run the project, follow these steps:
- Run "git clone ...".
- cd into the muscle_checker directory.
- Run "python3 -m venv venv" to create a new virtual environment to run the project inside.
- Run ". venv/bin/activate" to activate the new environment.
- Run "flask run --host=0.0.0.0" to start running the website.
- Go to "127.0.0.1:5000".
<br><br>

## **autogenerate_gym_calendar**<br>
**Date Started**: 2022-10-15<br>
**Date Completed**: 2022-10-28 (13 days)<br>
**Language**: React<br>
**Overview**: Investigate into whether it is possible to add events to your calendar given a form, just like barbers do. If you can, add a form to add your next week's gym schedule. There may be issues with adding to the correct calendar.
**Result**: I now have a autogenerate_script and calendar_index.html page to allow the user to create a gym schedule complete with links to add the events to their calendar. There are a couple of bugs still to fix.<br>
**Running the Project**: The website can be run simply by going to [http://192.168.1.109:5000/autogenerate_gym_calendar] and filling in the details.
<br><br>


muscle_checker/changelog.md
# muscle_checker
Last update: 2022-11-03 15:34
<br><br>

## Changelog for muscle_checker

1. Created the project and README.
2. Wrote all_muscles.json to contain all exercises.
3. Used a JSON Validator to check the formatting.
4. Created dataclasses for Group and Muscles.
5. Added a function to read from all_muscles.json.
6. Updated setup.py.
7. Got tox working.
    - Create a virtual environment - "python3 -m venv venv".
    - Activate it - ". venv/bin/activate". You can deactivate it with "deactivate".
    - pycodestyle "\" doesn't seem to be working. Just ignoring it.
8. Creates groups and muscles dicts to store all_muscles.
9. Wrote the read_exercises_from_text_file function.
10. Wrote the convert_exercises_list_to_dict function.
11. Started on find_missed_muscles.
12. Wrote suggested_exercises function.
13. Added a main method.
14. Added a way to take a parameter of the calendar file location. Run the script with: "python muscle_checker.py".
15. Added a GitLab action to learn how to do it using this <a href="https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions" target="_blank">guide</a>.
16. Made a new .github/workflows folder to hold workflows (actions) and made a Action that runs when muscle_checker changes.
17. Added linting to tox.ini. Run commands:
    - conda activate .venv/venv/activate
    - "tox" to run [testenv].
    - "tox -e lint" to run [testenv:lint].
    - "tox -e tests" to run [testenv:tests].
18. Finalised muscle_checker_action.yml. Only issue is that it doesn't fail if there are pycodestyle or pydocstyle issues.
19. Tested with a MR.
20. Cleaned up the changelog.
21. Fixed all the pycodestyle and pydocstyle errors.
22. Started the test_muscle_checker.py tests.
23. Finished all of the tests. Run then with "tox -e tests".
24. Worked on allowing several groups in one exercise list.
25. Started on the front-end app. 
    - Made app.py
    - Imports muscle_checker_script
    - Run it with " . venv/bin/activate"
    - "flask run --host=0.0.0.0"
26. Added an index.html and master.css.
27. Added content passed from python.
28. Moved the script runner into a run() method that can be used by app.py.
29. Made a web UI that prints the results with a preset location for the file.
30. Added an input form for the user to insert their workout.
31. Updated the script to take a string instead of a .txt file.
32. Fully working!
33. Removed unnecessary folders and added to .gitignore. Fixed the pycache files being added to pushes.
34. Cleaned up the README.
35. Looked into using autodocumentation (Sphinx-doc)
    - Updating the tox.ini file. "changedir = docs" creates a docs file.
    - Made a source file.
    - Made a conf.py file.
    - Starting again reading the tutorials.
36. Following the Sphinx documentation.
    - Activate the ve.
    - "pip install sphinx".
    - "sphinx-build --version".
    - "sphinx-quickstart docs" - Creates the files needed to get started - docs with build, source, make.bat and MakeFile inside. Source contains several files. make.bat and MakeFile are used for common Sphinx operations like rendering content.
    - "sphinx-build -b html docs/source/ docs/build/html" - Now have our first piece of documentation.
    - Updated index.rst.
    - Tutorial said to run "make html", but it is a .bat file so will not run.
    - "pip install furo". Already installed.
    - Made the [testenv:docs] from tox.ini.
    - Now can run it with "tox -e docs".
37. Added to "index.rst" to contain everything.
38. Played around with adding CSS.
39. Fixed linting in the pipleline. Now the GitHub runner fails if linting fails.


autogenerate_gym_calendar/autogenerate_script.py
"""This python script is used to generate the links required to make new calendar entries."""
from calendar import week
import urllib.parse
import random
from datetime import datetime

# Default used for testing.
default_form_details = {
    'week_monday': '2022-10-10',
    'days_selected': ['monday', 'tuesday', 'friday'],
    'gym': '1', 'runs': '1', 'other': '1',
    'specific_obligations': [['wednesday', 'kayaking']]
}


def run_autogenerate(form_details: dict = default_form_details) -> dict:
    """Generate all links inside of a dict to add events to the calendar.

    Args:
        form_details (dict, optional): A dictionary containing the form details from calendar_index.html. Defaults to default_form_details.

    Returns:
        dict: The calendar event links in a dict.
    """
    # Arrays to contain all links to calendar invites.
    monday_array = []
    tuesday_array = []
    wednesday_array = []
    thursday_array = []
    friday_array = []
    saturday_array = []
    sunday_array = []

    left = []  # Contains the number of each session type.
    left.append(int(form_details["gym"]))
    left.append(int(form_details["runs"]))
    left.append(int(form_details["other"]))

    total_sessions = left[0] + left[1]  # Totals the number of sessions.

    week_dates = calculate_dates(form_details["week_monday"])  # An array of dates for each weekday in the converted format.

    # Adds each specific obligation to the relevant arrays.
    for x in form_details["specific_obligations"]:

        sport = x[1]  # E.g. "kayaking"

        title_string = str(sport).capitalize() + ": 0 mins (00:00)"
        title_string = urllib.parse.quote(title_string, safe='')

        desc_string = urllib.parse.quote("Generated by autogenerate_script.", safe='')  # Converts the description to a URL-friendly format.

        event_sections = {
            "date": week_dates[0],  # Defaults to the date of the Monday.
            "title": title_string,  # E.g. "Kayaking: 0 mins (00:00)".
            "description": desc_string
        }

        if x[0] == "monday":
            event_sections.update({"date": week_dates[0]})
            monday_array.append([create_link(event_sections), str.capitalize(sport)])

        if x[0] == "tuesday":
            event_sections.update({"date": week_dates[1]})
            tuesday_array.append([create_link(event_sections), str.capitalize(sport)])

        if x[0] == "wednesday":
            event_sections.update({"date": week_dates[2]})
            wednesday_array.append([create_link(event_sections), str.capitalize(sport)])

        if x[0] == "thursday":
            event_sections.update({"date": week_dates[3]})
            thursday_array.append([create_link(event_sections), str.capitalize(sport)])

        if x[0] == "friday":
            event_sections.update({"date": week_dates[4]})
            friday_array.append([create_link(event_sections), str.capitalize(sport)])

        if x[0] == "saturday":
            event_sections.update({"date": week_dates[5]})
            saturday_array.append([create_link(event_sections), str.capitalize(sport)])

        if x[0] == "sunday":
            event_sections.update({"date": week_dates[6]})
            sunday_array.append([create_link(event_sections), str.capitalize(sport)])

    gym_sessions = ["Cardio", "Push", "Pull", "Legs"]  # Backwards so that they can be pop() off.

    days_remaining = form_details["days_selected"]

    # For each of the sessions that are being added.
    for x in range(0, total_sessions):

        # Refills gym_sessions if it is empty
        if len(gym_sessions) < 1:
            gym_sessions = ["Cardio", "Push", "Pull", "Legs"]

        # Ensures that a number isn't picked if it is 0. E.g. no runs left.
        not_allowed = True
        while not_allowed:
            chosen = random.randint(0, 1)
            # print("chosen: " + str(chosen))

            if chosen == 0 and left[0] > 0:
                not_allowed = False
                left[0] -= 1
            if chosen == 1 and left[1] > 0:
                not_allowed = False
                left[1] -= 1

        sessions = ["Gym", "Run"]

        if len(days_remaining) > 0:  # If not looped through all days.
            day = days_remaining.pop()
        else:  # Refills the days_remaining.
            days_remaining = form_details["days_selected"]

        gym_session = gym_sessions.pop()
        session_type = ""

        if chosen == 0:  # If a gym session.
            title_string = "Gym: " + gym_session + ", 0 mins (00:00)"
            session_type = "Gym (" + gym_session + ")"
        elif chosen == 1:  # If a run.
            title_string = "Ran 0.0K in 00:00"
            session_type = "Run"
        else:
            title_string = str(sessions[chosen]) + ": 0 mins (00:00)"
            session_type = str(sessions[chosen])
        title_string = urllib.parse.quote(title_string, safe='')

        desc_string = urllib.parse.quote("Generated by autogenerate_script.", safe='')

        event_sections = {
            "date": week_dates[0],  # Defaults to the Monday.
            "title": title_string,
            "description": desc_string
        }

        if day == "monday":
            event_sections.update({"date": week_dates[0]})
            monday_array.append([create_link(event_sections), session_type])

        if day == "tuesday":
            event_sections.update({"date": week_dates[1]})
            tuesday_array.append([create_link(event_sections), session_type])

        if day == "wednesday":
            event_sections.update({"date": week_dates[2]})
            wednesday_array.append([create_link(event_sections), session_type])

        if day == "thursday":
            event_sections.update({"date": week_dates[3]})
            thursday_array.append([create_link(event_sections), session_type])

        if day == "friday":
            event_sections.update({"date": week_dates[4]})
            friday_array.append([create_link(event_sections), session_type])

        if day == "saturday":
            event_sections.update({"date": week_dates[5]})
            saturday_array.append([create_link(event_sections), session_type])

        if day == "sunday":
            event_sections.update({"date": week_dates[6]})
            sunday_array.append([create_link(event_sections), session_type])

    link_dict = {
        "monday": monday_array,
        "tuesday": tuesday_array,
        "wednesday": wednesday_array,
        "thursday": thursday_array,
        "friday": friday_array,
        "saturday": saturday_array,
        "sunday": sunday_array
    }

    for day in link_dict:
        if link_dict[day] == []:
            link_dict[day] = [["-", "Rest"]]

    return link_dict


def create_link(event_sections: dict) -> str:
    """Given the event_section, creates the Goolge calendar link and returns it.

    Args:
        event_sections (dict): A dict containing the title, description and date of the event.

    Returns:
        str: The URL to the Google calendar event.
    """
    link_start = "https://calendar.google.com/calendar/render?action=TEMPLATE"

    dates = int(event_sections["date"])
    dated = str(dates) + "/" + str(dates + 1)
    link_date = f'&dates={dated}'

    # Description is optional, so have an option of it being blank.
    link_description = ""
    if event_sections["description"]:
        link_description = f'&details={event_sections["description"]}'

    # If there is no title, set it to "Title"
    try:
        link_title = f'&text={event_sections["title"]}'
    except KeyError:
        link_title = "&text=Title"

    # https://calendar.google.com/calendar/render?action=TEMPLATE&dates=20221025%2F20221026&details=The%20Desc&text=The%20Title
    link_full = link_start + link_date + link_description + link_title

    return link_full


def convert_date(date: str) -> str:
    """Convert date in the format of "2022-10-10" to "20221010".

    Args:
        date (str): The date in the original format.

    Returns:
        str: The date in the new format.
    """
    try:
        date_array = date.split("-")

        if len(date_array[0]) == 2:  # If the date given is in reverse order.
            date_array = date_array[::-1]

    except AttributeError:  # If the date provided is not a string return today instead.
        date = datetime.today().strftime('%Y-%m-%d')
        date_array = date.split("-")

    return "".join(date_array)


def calculate_dates(week_monday: str) -> list:
    """Convert the date from "2022-10-25" to "20221025" and makes sure that the day is in the correct month.

    Args:
        week_monday (str): The monday of the week.

    Returns:
        list: An array of all the dates converted.
    """
    week_monday_array = week_monday.split("-")  # Splits "2022-10-25" into ["2022", "10", "25"].
    year = week_monday_array[0]
    month = week_monday_array[1]
    day = week_monday_array[2]

    week_dates = [convert_date(week_monday)]  # Contains Monday.

    total_days = 31
    thirty_days = [4, 6, 9, 11]  # The dates with 30 days (i.e. April, June, etc)

    if month == 2:  # If the month is february.
        if int(year) % 4 == 0:  # If a leap year.
            total_days = 29
        else:
            total_days = 28

    elif month in thirty_days:  # If in a month with 30 days.
        total_days = 30

    for x in range(1, 7):  # 1, 2, 3, 4, 5, 6
        next_day = int(day) + x
        use_month = month

        if next_day > total_days:  # If it goes over the number of days in the month (e.g. 32)
            next_day = "0" + str(next_day - total_days)
            use_month = str(int(month) + 1)

        week_dates.append(str(year) + str(use_month) + str(next_day))

    return week_dates


print(run_autogenerate())
